### PR TITLE
feat: upvote streak

### PR DESCRIPTION
**Pull request by catngh**
**What does this PR do?**

-  Webhooks for external app to call to `/webhook/topgg` and `/webhook/discordbotlist`
- Upvote streak expire after 13 hours ( user can upvote every 12hr )
- API to get users upvote streak:
`curl --location --request GET 'http://localhost:8200/api/v1/users/upvote-streak?discord_id=319132138849173505'`
Response:
```
{
    "data": {
        "discord_id": "319132138849173505",
        "minutes_until_reset": 780,        -> this will be 0 if ran out of time
        "streak_count": 1,
        "total_count": 5,
        "last_streak_time": "2022-09-07T18:00:00Z"
    }
}
```

### CODE REVIEW

File: pkg/entities/user.go [hunk 1]
```diff
@@ -84,6 +84,19 @@ func (e *Entity) GetUserCurrentGMStreak(discordID, guildID string) (*model.Disco
 	return streak, http.StatusOK, nil
 }
 
+func (e *Entity) GetUserCurrentUpvoteStreak(discordID string) (*model.DiscordUserUpvoteStreak, int, error) {
+	streak, err := e.repo.DiscordUserUpvoteStreak.GetByDiscordID(discordID)
+	if err != nil && err != gorm.ErrRecordNotFound {
```
Comment (trkhoi) [inline pkg/entities/user.go:89]: log err

File: pkg/entities/user.go [hunk 2]
```diff
@@ -84,6 +84,19 @@ func (e *Entity) GetUserCurrentGMStreak(discordID, guildID string) (*model.Disco
...
+		return nil, http.StatusInternalServerError, fmt.Errorf("failed to get user's gm streak: %v", err)
+	}
+
+	if err == gorm.ErrRecordNotFound {
+		return nil, http.StatusBadRequest, fmt.Errorf("user has no gm streak")
```
Comment (trkhoi) [inline pkg/entities/user.go:94]: log err

File: pkg/entities/webhook.go [hunk 3]
```diff
@@ -407,3 +407,43 @@ func (e *Entity) BoostXPIncrease(message *discordgo.Message) (*response.HandleUs
...
+	}
+
+	if err == gorm.ErrRecordNotFound {
+		err = e.repo.DiscordUserUpvoteStreak.UpsertOne(model.DiscordUserUpvoteStreak{
+			DiscordID:      userID,
+			StreakCount:    1,
+			TotalCount:     1,
+			LastStreakDate: chatDate,
+		})
+		if err != nil {
```
Comment (trkhoi) [inline pkg/entities/webhook.go:426]: log err

File: pkg/entities/webhook.go [hunk 4]
```diff
@@ -407,3 +407,43 @@ func (e *Entity) BoostXPIncrease(message *discordgo.Message) (*response.HandleUs
 
 	return resp, nil
 }
+
+func (e *Entity) WebhookUpvoteStreak(userID string) error {
+	sentAt := time.Now()
+	chatDate := time.Date(sentAt.Year(), sentAt.Month(), sentAt.Day(), sentAt.Hour(), 0, 0, 0, time.UTC)
+	streak, err := e.repo.DiscordUserUpvoteStreak.GetByDiscordID(userID)
+	if err != nil && err != gorm.ErrRecordNotFound {
```
Comment (trkhoi) [inline pkg/entities/webhook.go:415]: log err

File: pkg/entities/webhook.go [hunk 5]
```diff
@@ -407,3 +407,43 @@ func (e *Entity) BoostXPIncrease(message *discordgo.Message) (*response.HandleUs
...
+	case chatDate.Equal(nextStreakDate):
+		streak.StreakCount++
+	case chatDate.After(nextStreakDate):
+		streak.StreakCount = 1
+	}
+	streak.LastStreakDate = chatDate
+	streak.TotalCount++
+
+	if err := e.repo.DiscordUserUpvoteStreak.UpsertOne(*streak); err != nil {
+		return fmt.Errorf("failed to update user upvote streak: %v", err)
```
Comment (trkhoi) [inline pkg/entities/webhook.go:446]: log err

**Review by trkhoi — CHANGES_REQUESTED** [review]

**Review by trkhoi — APPROVED** [review]

**Review by tuanddd — APPROVED** [review]

**Pull request merged**